File: frontend/src/utils/geo.ts
const EARTH_RADIUS_KM = 6371

export interface LatLng {
  lat: number
  lng: number
}

function toRadians(deg: number): number {
  return (deg * Math.PI) / 180
}

export function distanceKm(a: LatLng, b: LatLng): number {
  const dLat = toRadians(b.lat - a.lat)
  const dLng = toRadians(b.lng - a.lng)
  const lat1 = toRadians(a.lat)
  const lat2 = toRadians(b.lat)

  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2

  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(h))
}

export function isWithinRevealRadius(
  lat: number,
  lng: number,
  visited: LatLng[],
  radiusKm: number
): boolean {
  return visited.some(v => distanceKm({ lat, lng }, v) <= radiusKm)
}

File: backend/services/geo.go
package services

import "math"

const earthRadiusMiles = 3958.8

// HaversineMiles returns the great-circle distance in miles between two
// lat/lng points, using the haversine formula. Shared by trail proximity
// matching (POST /trail) and, later, travel-time/distance-sort features —
// written once here so both consume the same math.
func HaversineMiles(lat1, lng1, lat2, lng2 float64) float64 {
	toRad := func(deg float64) float64 { return deg * math.Pi / 180 }

	dLat := toRad(lat2 - lat1)
	dLng := toRad(lng2 - lng1)
	rlat1 := toRad(lat1)
	rlat2 := toRad(lat2)

	h := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(rlat1)*math.Cos(rlat2)*math.Sin(dLng/2)*math.Sin(dLng/2)

	return 2 * earthRadiusMiles * math.Asin(math.Sqrt(h))
}
